crates/circuit/src/dag_node.rs

```diff
@@ -0,0 +1,329 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::circuit_instruction::CircuitInstruction;
+use pyo3::prelude::*;
+use pyo3::sync::GILOnceCell;
+use pyo3::types::{PyDict, PyList, PySequence, PyString, PyTuple};
+use pyo3::{intern, PyObject, PyResult};
+
+static SEMANTIC_EQ_PYTHON: GILOnceCell<PyResult<PyObject>> = GILOnceCell::new();
```

Comment:
This kind of violates the idea of Python's virtual dispatch for the method - the method exists on a module, but it's no longer possible to meaningfully monkey-patch it in a testing context or anything like that.

That's probably not so important, but is there any meaningful performance hit from just re-importing it from a suitable module each time it's used?

Reply:
This specific concern at least should be resolved with 041f3419b06b6f3a8c1a9cb0cc1db0021d357d7c.